Add hubble_relay_pool_peer_connection_status metric

siwiutki opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

This change adds a new gauge metric to hubble-relay measuring the connectiion status to all peers. Metric keeps track of number of peers for each possible connectiion status

The current set of metrics is not enough to accurately measure the availability of hubble-relay. They measure the status of gRPC calls, but, for instance, in case all peers are unreachable when GetFlows is called, even though gRPC call will succeed and return "OK" status, the response will come with no flows gathered, rendering it useless. This new metric is introduced to cover such cases.

Fixes #27890

```release-note
Added hubble_relay_pool_peer_connection_status metric for measuring the connection status of all peers. Metric keeps track of number of peers for each possible connectiion status.
```

marqc commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -356,3 +375,44 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
+			s.opts.log.WithField("address", p.Address).Infof(
+				"No connection to peer %s, skipping", p.Name,
+			)
+			s.peers.ReportOffline(p.Name)
```
This looks like changing some state in the peers view. As this is only used in observability/metrics code it should not affect any state and work as read-only code.

siwiutki replied:
I agree with your other comment to move this to PeerManager. Since in that component there is already a periodic check for offline peers, I think we can remove it from reporting status as well.

marqc commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -356,3 +375,44 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
+	for _, p := range peers {
+		if !isAvailable(p.Conn) {
+			numUnavailablePeers++
+			s.opts.log.WithField("address", p.Address).Infof(
```
nit: If you just report metrics with peers status, i would not log unavailable nodes here (or change log level to debug).

siwiutki replied:
With heartbeat run every 5 seconds, this might actually be spamming logs very often and this info will be available through metrics anyway. I'm rather with you on removing this log instead of changing to debug.

marqc commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -356,3 +375,44 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
+	var numConnectedPeers, numUnavailablePeers uint32
+	peers := s.peers.List()
+	for _, p := range peers {
+		if !isAvailable(p.Conn) {
```
the `isAvailable` categorizes grpc connection (between relay and cilium-agent) states into 2 categories. The "Ready" and "Idle" states are  categorized as available. All other states are categorized as unavailable.

When debugging peer connectivity issues I believe it can be more useful to report the actual grpc connection state as defined in https://github.com/cilium/cilium/blob/main/vendor/google.golang.org/grpc/connectivity/connectivity.go#L31

siwiutki replied:
Sure.

marqc commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -28,6 +30,8 @@ import (
 // Reporting up to 10 unavailable nodes is probably reasonable.
 const numUnavailableNodesReportMax = 10
 
+const heartbeatInterval = time.Minute
```
nit: This seems to be quite a lightweight code path to execute (check and report status). I believe it makes sense to call this method more often, i.e. every 5 seconds.

siwiutki replied:
Sure.

marqc reviewed (commented):
Overall a good job. Thanks for implementing this. I have one suggestion to report real grpc peer connection status instead of pre-aggregated as available/non-available.

siwiutki committed:
Add hubble_relay_pool_peer_connection_status metric

This change adds a new gauge metric to hubble-relay measuring the connectiion status to all peers. Metric keeps track of number of peers for each possible connectiion status

The current set of metrics is not enough to accurately measure the availability of hubble-relay. They measure the status of gRPC calls, but, for instance, in case all peers are unreachable when GetFlows is called, even though gRPC call will succeed and return "OK" status, the response will come with no flows gathered, rendering it useless. This new metric is introduced to cover such cases.

Signed-off-by: Michal Siwinski <siwy@google.com>

siwiutki reviewed (commented):
I agree that this code fits better into the PeerManager code. Moved it there already.

Unit tests that I wrote for PeerManager utilise `time.Sleep` to test the actual code that exports this metric. I don't fully like it, but I don't see time being mocked anywhere in the current unit tests, so I didn't try to reinvent the wheel here. Timer used in PeerManager could be mocked and usually https://github.com/benbjohnson/clock library is a good pick for that.

marqc commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -186,6 +194,41 @@ func (m *PeerManager) manageConnections() {
+			}
+			m.mu.RUnlock()
+
+			m.mu.Lock()
```
nit: this lock is unnecessary, as the `ObservePeerConnectionStatus` only does the `Set` method on `prometheus GaugeVec`, and prometheus client is already thread-safe.

siwiutki replied:
The lock was definitely needed, because `ObservePeerConnectionStatus` was not thread-safe - even though setting Prometheus gauge for a particular label was thread-safe, this function was calling `Set` multiple times and so having states {"IDLE": 3, "READY": 1} and {"IDLE": 1, "READY": 3} being observed at the same time, you could end up with states {"IDLE": 3, "READY": 3} being displayed through gauge metric.

Now though I just made the `ObservePeerConnectionStatus` thread-safe itself, so that the lock used in metrics won't block other PeerManager operations. Useful nit to that, so thanks.

marqc replied:
I like your solution with having a separate mutex for the metric gauge setter. Thanks.

marqc reviewed (approved):
LGTM! Great job! Thanks for contributing!

kaworu commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -186,6 +194,38 @@ func (m *PeerManager) manageConnections() {
+				// which happens to be a perfect initial value
+				num, _ := connStates[state]
+				num++
+				connStates[state] = num
```
nit: IMO it's fine to "silently" rely on zero values:
```suggestion
				connStates[state] = connStates[state] + 1
```

siwiutki replied:
Sure, applied your suggestion.

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -704,6 +709,176 @@ func TestPeerManager(t *testing.T) {
+			assert.NoError(t, err)
+			mgr.Start()
+			// sleep so that metrics will be emitted at least once
+			time.Sleep(3 * time.Second)
```
I'm a little worried to rely on sleeping here, as it is a common source of flaky tests. Is there a way to loop until the metric has changed once instead?

siwiutki replied:
Since you're using testify framework for tests, `assert.Eventually` is IMO a clean approach to what you're suggesting. And I assume that loop would need a timeout anyway, so it fits very well here. Changed the test to use this instead of that pesky sleep.

glrf commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -704,6 +709,176 @@ func TestPeerManager(t *testing.T) {
+				"IDLE":              2,
+				"TRANSIENT_FAILURE": 1,
+				"SHUTDOWN":          1,
+				"NIL_CONNECTION":    0,
```
nit: It might be a good idea to add a test case that actually returns a `NIL_CONNECTION`.  Currently we could just remove all of the `nilConnNum` counting logic and the tests will still pass.

siwiutki replied:
Sure, added a test case for that.

kaworu reviewed (commented):
Hi @siwiutki :wave: and thanks for the PR!

Couple of comment, but overall LGTM. Agree with @glrf that testing `NIL_CONNECTION` would be an improvement.

kaworu reviewed (approved):
Thanks @siwiutki!

siwiutki reviewed (commented):
The second commit was a merge commit to be in sync with current main branch. Synced it with rebase instead of merge, so should be fine with 1 commit now.

rolinh reviewed (approved):
Awesome work, lgtm!
